core/engine/src/builtins/date/utils.rs

```diff
@@ -1,202 +1,730 @@
-    // 1𝔽; but if this is not possible (because some argument is out of range), return NaN.
-    let month = usize::try_from(month).expect("month must be between 0 and 11 at this point");
-
-    let mut day = day_from_year(year);
-
-    // Consider leap years when calculating the cumulative days added to the year from the input month
-    if (year % 4 != 0) || (year % 100 == 0 && year % 400 != 0) {
-        day += [0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334][month];
-    } else {
-        day += [0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335][month];
-    }
+    // 7. Let mn be 𝔽(ℝ(m) modulo 12).
+    let mn = modulo(m, 12.0) as u8;
+
+    // 8. Find a finite time value t such that YearFromTime(t) is ym, MonthFromTime(t) is mn,
+    //    and DateFromTime(t) is 1𝔽;
+    //    but if this is not possible (because some argument is out of range), return NaN.
+    let rest = if mn > 1 { 1.0 } else { 0.0 };
+    let days_within_year_to_end_of_month = match mn {
+        0 => 0.0,
+        1 => 31.0,
+        2 => 59.0,
+        3 => 90.0,
+        4 => 120.0,
+        5 => 151.0,
+        6 => 181.0,
+        7 => 212.0,
+        8 => 243.0,
+        9 => 273.0,
+        10 => 304.0,
+        11 => 334.0,
+        12 => 365.0,
+        _ => unreachable!(),
+    };
+    let t =
+        (day_from_year(ym + rest) - 365.0 * rest + days_within_year_to_end_of_month) * MS_PER_DAY;
 
     // 9. Return Day(t) + dt - 1𝔽.
-    (day - 1).checked_add(date)
+    day(t) + dt - 1.0
 }
 
-/// Abstract operation [`MakeDate`][spec].
+/// Abstract operation `MakeDate ( day, time )`
 ///
-/// [spec]: https://tc39.es/ecma262/multipage/numbers-and-dates.html#sec-makedate
-pub(crate) fn make_date(day: i64, time: i64) -> Option<i64> {
+/// More info:
+/// - [ECMAScript reference][spec]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-makedate
+pub(super) fn make_date(day: f64, time: f64) -> f64 {
     // 1. If day is not finite or time is not finite, return NaN.
+    if !day.is_finite() || !time.is_finite() {
+        return f64::NAN;
+    }
+
     // 2. Let tv be day × msPerDay + time.
+    let tv = day * MS_PER_DAY + time;
+
     // 3. If tv is not finite, return NaN.
+    if !tv.is_finite() {
+        return f64::NAN;
+    }
+
     // 4. Return tv.
-    day.checked_mul(MILLIS_PER_DAY)?.checked_add(time)
+    tv
 }
 
-/// Abstract operation [`TimeClip`][spec]
-/// Returns the timestamp (number of milliseconds) if it is in the expected range.
-/// Otherwise, returns `None`.
+/// Abstract operation `MakeFullYear ( year )`
+///
+/// More info:
+/// - [ECMAScript reference][spec]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-makefullyear
+pub(super) fn make_full_year(year: f64) -> f64 {
+    // 1. If year is NaN, return NaN.
+    if year.is_nan() {
+        return f64::NAN;
+    }
+
+    // 2. Let truncated be ! ToIntegerOrInfinity(year).
+    let truncated = IntegerOrInfinity::from(year);
```

Comment:
thought: is `IntegerOrInfinity` needed in this instance if we've already accounted for NaN?

Reply:
IIRC, `IntegerOrInfinity` also rounds to an integer, so I think it's pretty much needed here.

Reply:
Yeah, I think it's just truncating.

So I was just wondering whether this step could just be `year.trunc()` in this particular instance.

Reply:
Hmmm that is true... `IntegerOrInfinity` is mostly used to convert from `f64` to an integer, or handle the non-finite edge cases, but here it's only used for the rounding...

Reply:
It can be left as is and merged. It was just something I when I was reviewing.

I think this could mostly be expressed as the below.

```rust
// Return non-finite cases
if !year.is_finite() {
    return year
}
// truncate
let trunc = year.trunc();

// Check if value is in range.
if (0.0..=99.0).contains(&year) {
    return year 1900.0 + year
}

year

```

Reply:
I think I wanted to leave this for the optimizations later. For example in this case, every result of `make_full_year` gets passed to `make_day` afterwards. So there is definitley some potential to optimize this even across spec method boundaries.